component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java

```diff
@@ -0,0 +1,272 @@
+            if (StringUtils.isEmpty(tokenEndPointAlias)) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Token Endpoint alias of the local Identity " +
+                        "Provider has not been configured for " + identityProvider.getIdentityProviderName());
+            }
+
+            audienceFound = validateAudience(audiences, tokenEndPointAlias, requested_audience);
+            if (!audienceFound) {
+                handleException(TokenExchangeConstants.INVALID_TARGET, "Invalid audience values provided");
+            }
+            TokenExchangeUtils.checkExpirationTime(expirationTime, currentTimeInMillis, timeStampSkewMillis);
+            if (notBeforeTime != null) {
+                TokenExchangeUtils.checkNotBeforeTime(notBeforeTime, currentTimeInMillis, timeStampSkewMillis);
+            } else {
+                log.debug("Not Before Time(nbf) not found in JWT. Continuing Validation");
+            }
+            if (issuedAtTime == null) {
+                log.debug("Issued At Time(iat) not found in JWT. Continuing Validation");
+            } else if (!validateIAT) {
+                log.debug("Issued At Time (iat) validation is disabled for the JWT");
+            } else {
+                TokenExchangeUtils.checkValidityOfTheToken(issuedAtTime, currentTimeInMillis, timeStampSkewMillis,
+                        validityPeriod);
+            }
+            boolean customClaimsValidated = validateCustomClaims(customClaims);
+            if (!customClaimsValidated) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Custom Claims in the JWT were invalid");
+            }
+            log.debug("Subject JWT Token was validated successfully");
+            if (OAuth2Util.isOIDCAuthzRequest(tokReqMsgCtx.getScope())) {
+                TokenExchangeUtils.handleCustomClaims(tokReqMsgCtx, customClaims, identityProvider, tenantDomain,
+                        registeredClaimNames);
+            }
+        } else {
+            handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Unsupported Requested Token Type : " +
+                    requested_token_type + " provided");
+        }
+        return true;
+    }
+
+    @Override
+    public OAuth2AccessTokenRespDTO issue(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+        OAuth2AccessTokenRespDTO tokenRespDTO = super.issue(tokReqMsgCtx);
+        AuthenticatedUser user = tokReqMsgCtx.getAuthorizedUser();
+        Map<ClaimMapping, String> userAttributes = user.getUserAttributes();
+        if (MapUtils.isNotEmpty(userAttributes)) {
+            ClaimsUtil.addUserAttributesToCache(tokenRespDTO, tokReqMsgCtx, userAttributes);
+        }
+        tokenRespDTO.addParameter(TokenExchangeConstants.ISSUED_TOKEN_TYPE, requested_token_type);
+        return tokenRespDTO;
+    }
+
+    @Override
+    public boolean issueRefreshToken() throws IdentityOAuth2Exception {
+        return OAuthServerConfiguration.getInstance()
```

Comment:
Add a blank line here.

Reply:
Let's add method comments to all the public methods.

Reply:
Fixed in 8693d5454cb05015e9d9614c56a1d09bdf8a8997